packages/runtime/runtime-utils/src/summarizerNode/summarizerNodeWithGc.ts

```diff
@@ -114,6 +158,94 @@ export class SummarizerNodeWithGC extends SummarizerNode implements IRootSummari
         return gcData;
     }
 
+    /**
+     * Called during the start of a summary. Updates the work-in-progress used routes.
+     */
+    public startSummary(referenceSequenceNumber: number, summaryLogger: ITelemetryLogger) {
+        assert(this.wipSerializedUsedRoutes === undefined, "wip routes should not be set yet in startSummary");
+        this.wipSerializedUsedRoutes = JSON.stringify(this.usedRoutes);
+        super.startSummary(referenceSequenceNumber, summaryLogger);
+    }
+
+    /**
+     * Called after summary has been uploaded to the server. Add the work-in-progress state to the pending
+     * summary queue. We track this until we get an ack from the server for this summary.
+     */
+    protected completeSummaryCore(
+        proposalHandle: string,
+        parentPath: EscapedPath | undefined,
+        parentSkipRecursion: boolean,
+    ) {
+        const wipSerializedUsedRoutes = this.wipSerializedUsedRoutes;
+        assert(wipSerializedUsedRoutes !== undefined, "wip routes should have been set in startSummary");
+
+        super.completeSummaryCore(proposalHandle, parentPath, parentSkipRecursion);
+
+        const summaryNode = this.pendingSummaries.get(proposalHandle);
+        if (summaryNode !== undefined) {
+            const summaryNodeWithGC = new SummaryNodeWithGC(wipSerializedUsedRoutes, summaryNode);
+            this.pendingSummaries.set(proposalHandle, summaryNodeWithGC);
+        }
+    }
+
+    /**
+     * Clears the work-in-progress state.
+     */
+    public clearSummary() {
+        this.wipSerializedUsedRoutes = undefined;
+        super.clearSummary();
+    }
+
+    /**
+     * Called when we get an ack from the server for a summary we sent. Update the reference state of this node
+     * from the state in the pending summary queue.
+     */
+    protected refreshLatestSummaryFromPending(
+        proposalHandle: string,
+        referenceSequenceNumber: number,
+    ): void {
+        const summaryNode = this.pendingSummaries.get(proposalHandle) as SummaryNodeWithGC;
+        if (summaryNode !== undefined) {
+            this.referenceUsedRoutes = JSON.parse(summaryNode.serializedUsedRoutes);
+        }
+
+        return super.refreshLatestSummaryFromPending(proposalHandle, referenceSequenceNumber);
+    }
+
+    /**
+     * Called when we need to upload the reference state from the given summary. Read the GC blob and get the state
+     * to upload from it.
+     */
+    protected async refreshLatestSummaryFromSnapshot(
+        referenceSequenceNumber: number,
+        snapshotTree: ISnapshotTree,
+        basePath: EscapedPath | undefined,
+        localPath: EscapedPath,
+        correlatedSummaryLogger: ITelemetryLogger,
+        readAndParseBlob: ReadAndParseBlob,
+    ): Promise<void> {
+        const gcDetailsHash = snapshotTree.blobs[gcBlobKey];
```

Comment:
nit: I don't think this is always a hash(?) or if so that would be an implementation detail leaking here to the variable name.

Reply:
good point. Renamed it to gcDetailsBlob